### PR TITLE
fix: ensure completion script uses valid function name

### PR DESCRIPTION
**Pull request by davidmurdoch**
Fixes https://github.com/yargs/yargs/issues/2387

### CODE REVIEW

**Comment by davidmurdoch** [comment]
Anything else you need for this?

File: lib/completion.ts [hunk 1]
```diff
@@ -349,7 +349,15 @@ export class Completion implements CompletionInstance {
     let script = this.zshShell
       ? templates.completionZshTemplate
       : templates.completionShTemplate;
-    const name = this.shim.path.basename($0);
+    let name = this.shim.path.basename($0);
+    // Santize `name` to make sure it is a valid bash function name
+    // first, check if the first character is an alphabetic character or underscore
```
Comment (shadowspawn) [inline lib/completion.ts:354]: I think the completion template might be putting an underscore on the front already so this step is not needed? (Based on code inspection, not from running code.)

https://github.com/yargs/yargs/blob/0c95f9c79e1810cf9c8964fbf7d139009412f7e7/lib/completion-templates.ts#L8

File: lib/completion.ts [hunk 2]
```diff
@@ -349,7 +349,15 @@ export class Completion implements CompletionInstance {
...
+    if (!str.match(/^[A-Za-z_]/)) {
+      // If not, prepend name with an underscore
+      name = '_' + name;
+    }
+    // Then, replace any characters that are not alphanumeric or underscore with an underscore
+    name = name.replace(/[^A-Za-z0-9_]/g, '_');
```
Comment (davidmurdoch) [inline lib/completion.ts:360]:
```suggestion
    // Santize `name` to make sure it is a valid bash function name by
    // replacing any characters that are not alphanumeric or underscore with an underscore
    name = name.replace(/[^A-Za-z0-9_]/g, '_');
```

File: lib/completion.ts (hunk 1 above)
Comment (davidmurdoch) [reply lib/completion.ts]: ah, good catch

File: test/completion.cjs [hunk 3]
```diff
@@ -608,6 +608,14 @@ describe('Completion', () => {
       r.logs[0].should.match(/ndm flintlock >>/);
     });
 
+    it('santizes scriptName ($0) for use in a bash function', () => {
+      const r = checkUsage(
+        () => yargs([]).scriptName("1 test ing").showCompletionScript()
```
Comment (davidmurdoch) [inline test/completion.cjs:613]:
```suggestion
        () => yargs([]).scriptName('1 test ing').showCompletionScript()
```

**Comment by davidmurdoch** [comment]
Is there anything I can do to push this over the line?

**Comment by shadowspawn** [comment]
Ok, my main problem is I am dubious this is useful in practice. Working with command names with spaces in them is pretty clunky and I have never seen it being used. I need convincing that users will benefit from this before I give it a positive review.  (I don't think completion will work through a yarn run script, but haven't tried that scenario mentioned in the issue.)

My secondary concern is whether it even works. I had another look. The current PR completion does not seem to be functional for a script name with a space as the `compdef` line at the end is adding completion for `foo_bar`.

```
% 'foo bar' completion        
#compdef foo_bar
###-begin-foo_bar-completions-###
#
# yargs command completion script
#
# Installation: foo bar completion >> ~/.zshrc
#    or foo bar completion >> ~/.zprofile on OSX.
#
_foo_bar_yargs_completions()
{
  local reply
  local si=$IFS
  IFS=$'
' reply=($(COMP_CWORD="$((CURRENT-1))" COMP_LINE="$BUFFER" COMP_POINT="$CURSOR" foo bar --get-yargs-completions "${words[@]}"))
  IFS=$si
  _describe 'values' reply
}
compdef _foo_bar_yargs_completions foo_bar
###-end-foo_bar-completions-###
```

I wrote out the completion and edited it to have:
```
compdef _foo_bar_yargs_completions 'foo bar'
```

And could then get completion for 

```
% 'foo bar' --<tab>
--boolean  --string  --                                                                                                                                              
--help               -- Show help                                                                                                                                    
--version            -- Show version number                                                                                                                          
```

**Comment by davidmurdoch** [comment]
> I have never seen it being used.

Its in every npm script that ends up calling out out to yargs.

And I'm using it in the MetaMask extension, er, I want to use it in MetaMask but can't without quite a bit of customization because the generated completion script isn't valid.

I've been editing the minified code, so sorry my PR doesn't even work 😢 

Looks like zhs just needs single quotes around commands with spaces. But bash needs a bit more work, as it doesn't work even when quoted. The bash version would be something like:

```
_yarn_webpack_yargs_completions()
{
    local cur prev command_line

    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    command_line="${COMP_WORDS[*]}"

    # Check if the command line specifically starts with 'yarn webpack'
    if [[ ${command_line} == yarn\ webpack* ]]; then
        # Your completion logic goes here
        # Ensure we are adjusting the argument list appropriately
        local adjusted_args=("${COMP_WORDS[@]:2}") # Skip 'yarn' and 'webpack'
        local type_list

        # Simulate the command as if it started with 'webpack' for yargs
        type_list=$(yarn webpack --get-yargs-completions "${adjusted_args[@]}")
        COMPREPLY=( $(compgen -W "${type_list}" -- ${cur}) )
        return 0
    fi
}

complete -F _yarn_webpack_yargs_completions yarn

**Comment by shadowspawn** [comment]
> It's in every npm script that ends up calling out out to yargs.

This might be where I am missing something. How does this involve a command name with a space in it? I'll explain what I am thinking.

So in my head, I have an npm run-script called `foo`. The run-script uses (say) a command named `bar`. The user types `npm run foo -- --compress` and it invokes `bar --compress`. You suggested the author might specify `.scriptName('npm run foo')` to make the help reflect the main way it is called. I can see that.

Where does completion come into it? I don't think the user will be able to get completion on `npm run foo -- --f<tab>`. What is the command that the user types that gets completion? 

(One step simpler with yarn if can omit the `run`, but I think otherwise similar.)

**Comment by shadowspawn** [comment]
> And I'm using it in the MetaMask extension

This sound like you do have a concrete plan! What does the help show, and what is user typing on command-line when completion works? Or is the completion used in a non-interactive way by some higher level tooling?

**Comment by shadowspawn** [comment]
(I am not expert in completion handlers, and my mental model of one handler per command might be preventing me seeing the goal.)

**Comment by shadowspawn** [comment]
(Oh, multi-word completion handlers, that rings a vague bell...)

**Comment by davidmurdoch** [comment]
Closing, superseded by PR #2422

**Pull request closed without merging**